### PR TITLE
Improve extractor logging and a minor readability fix

### PR DESCRIPTION
**Pull request by sauyon**
These changes are mostly changes that were for Kubernetes.

 - Package directory as a wanted root
  
   This makes the extractor actually extract vendor packages when they're explicitly passed to a build command; before the wanted root would be the mod directory, and so the all the files would because the relative path to the root includes `vendor`. I'm not 100% sure we should do this, but that was my intention when I made the original changes, so I've changed this.

 - Enable Go modules while determining module directory

   This force enables modules while running `go list -f {{.Module}} ...` so that even if `GO111MODULE=off` is set manually, we can still detect module directory. This is done by Kubernetes, and the change increases the coverage of the database significantly.

 - Improve extractor logging

   Added a bunch of miscellaneous logging. I've discovered that a lot of the time spent extracting is actually spent in the single-threaded bit where we walk dependencies and extract types; ideally we'd be able to parallelize that a bit with a topological sort or something similar.

### CODE REVIEW

File: extractor/extractor.go [hunk 1]
```diff
@@ -32,15 +32,10 @@ func Extract(patterns []string) error {
...
 // ExtractWithFlags extracts the packages specified by the given patterns and build flags
 func ExtractWithFlags(buildFlags []string, patterns []string) error {
-	cfg := &packages.Config{
-		Mode: packages.NeedName | packages.NeedFiles |
-			packages.NeedCompiledGoFiles |
-			packages.NeedImports | packages.NeedDeps |
-			packages.NeedTypes | packages.NeedTypesSizes |
-			packages.NeedTypesInfo | packages.NeedSyntax,
-		BuildFlags: buildFlags,
+	modEnabled := os.Getenv("GO111MODULE") == "off"
```
Comment (smowton) [inline extractor/extractor.go:35]: This is backwards

File: extractor/extractor.go [hunk 2]
```diff
@@ -49,15 +44,28 @@ func ExtractWithFlags(buildFlags []string, patterns []string) error {
 		}
 	}
 
+	log.Println("Running Go frontend.")
```
Comment (smowton) [inline extractor/extractor.go:47]: Make more precise: `Running packages.Load`

File: extractor/util/util.go [hunk 3]
```diff
@@ -28,9 +28,14 @@ func Getenv(key string, aliases ...string) string {
...
 // package `pkgpath`.
 func runGoList(format string, pkgpath string, flags ...string) (string, error) {
+	return runGoListWithEnv(format, pkgpath, nil, flags...)
+}
+
+func runGoListWithEnv(format string, pkgpath string, additionalEnv []string, flags ...string) (string, error) {
 	args := append([]string{"list", "-e", "-f", format}, flags...)
 	args = append(args, pkgpath)
 	cmd := exec.Command("go", args...)
+	cmd.Env = append(os.Environ(), additionalEnv...)
```
Comment (smowton) [inline extractor/util/util.go:38]: To check: is `append` sufficient if the var is already present in the envrionment?

File: extractor/util/util.go [hunk 4]
```diff
@@ -48,13 +53,13 @@ func runGoList(format string, pkgpath string, flags ...string) (string, error) {
 // GetModDir gets the absolute directory of the module containing the package with path
 // `pkgpath`. It passes the `go list` the flags specified by `flags`.
 func GetModDir(pkgpath string, flags ...string) string {
-	mod, err := runGoList("{{.Module}}", pkgpath, flags...)
+	mod, err := runGoListWithEnv("{{.Module}}", pkgpath, []string{"GO111MODULE=on"}, flags...)
```
Comment (smowton) [inline extractor/util/util.go:56]: Comment here on why you force-enable module mode

**Review by smowton — CHANGES_REQUESTED** [review]
Please add some tests; what's the eval status here?

File: extractor/util/util.go (hunk 3 above)
Comment (sauyon) [reply extractor/util/util.go:38]: Yep

**Comment by sauyon** [comment]
I don't think I can add tests under our current framework because of the limitation on `qltest` where we can't specify build commands.

I haven't actually run any evaluations on this. Maybe I should have included this in the build tracing PR, since it's all only relevant to traced builds (I think).

**Comment by smowton** [comment]
Could you also do an lgtm-distribution test to check whether this derails any existing projects?

**Comment by smowton** [comment]
Tests: I suppose I was imagining some project in `extractor-tests` that extracts usefully with these changes but not without them, something to ensure this is providing the value it's supposed to

**Comment by sauyon** [comment]
> Tests: I suppose I was imagining some project in `extractor-tests` that extracts usefully with these changes but not without them, something to ensure this is providing the value it's supposed to

The issue is that I don't think this should change anything for non-traced builds, which I believe is the only thing we can test at the moment (@max-schaefer?).

**Comment by sauyon** [comment]
Actually, I'll just fold everything but the logging changes into the build-tracing PR; I think that makes more sense.

**Review by smowton — APPROVED** [review]
LGTM

File: extractor/extractor.go [hunk 5]
```diff
@@ -72,6 +81,8 @@ func ExtractWithFlags(buildFlags []string, patterns []string) error {
 	packages.Visit(pkgs, func(pkg *packages.Package) bool {
 		return true
 	}, func(pkg *packages.Package) {
+		log.Printf("Processing package %s.", pkg.PkgPath)
```
Comment (max-schaefer) [inline extractor/extractor.go:83]: Does this (and the other uses of `Printf`) want a final `\n`, or do we get that from somewhere else?

File: extractor/extractor.go [hunk 6]
```diff
@@ -163,10 +181,15 @@ func ExtractWithFlags(buildFlags []string, patterns []string) error {
 			extractPackage(pkg, &wg, goroutineSem, fdSem)
 			return
 		}
+
+		log.Printf("Skipping dependency package %s", pkg.PkgPath)
```
Comment (max-schaefer) [inline extractor/extractor.go:184]:
```suggestion
		log.Printf("Skipping dependency package %s.", pkg.PkgPath)
```

**Review by max-schaefer — CHANGES_REQUESTED** [review]

File: extractor/extractor.go (hunk 5 above)
Comment (sauyon) [reply extractor/extractor.go:83]: For some reason `log.Printf` adds a newline at the end.

**Comment by sauyon** [comment]
Made the changes and rebased.

File: extractor/extractor.go (hunk 5 above)
Comment (max-schaefer) [reply extractor/extractor.go:83]: :+1:

**Review by max-schaefer — APPROVED** [review]

**Pull request merged**